bee-api/bee-rest-api/src/endpoints/routes/api/v2/add_peer.rs

```diff
@@ -1,87 +1,55 @@
-// Copyright 2020-2021 IOTA Stiftung
+// Copyright 2020-2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use std::net::IpAddr;
-
-use bee_gossip::{Command::AddPeer, Multiaddr, NetworkCommandSender, PeerId, PeerRelation, Protocol};
-use bee_protocol::workers::PeerManager;
-use bee_runtime::resource::ResourceHandle;
-use serde_json::Value as JsonValue;
-use warp::{filters::BoxedFilter, http::StatusCode, reject, Filter, Rejection, Reply};
+use axum::{
+    extract::{Extension, Json},
+    response::IntoResponse,
+    routing::post,
+    Router,
+};
+use bee_gossip::{Command::AddPeer, Multiaddr, PeerId, PeerRelation, Protocol};
+use log::error;
+use serde_json::Value;
 
 use crate::{
-    endpoints::{
-        config::ROUTE_ADD_PEER,
-        filters::{with_network_command_sender, with_peer_manager},
-        permission::has_permission,
-        rejection::CustomRejection,
-    },
+    endpoints::{error::ApiError, storage::StorageBackend, ApiArgsFullNode},
     types::{
         dtos::{PeerDto, RelationDto},
         responses::AddPeerResponse,
     },
 };
 
-fn path() -> impl Filter<Extract = (), Error = warp::Rejection> + Clone {
-    super::path().and(warp::path("peers")).and(warp::path::end())
-}
-
-pub(crate) fn filter(
-    public_routes: Box<[String]>,
-    allowed_ips: Box<[IpAddr]>,
-    peer_manager: ResourceHandle<PeerManager>,
-    network_command_sender: ResourceHandle<NetworkCommandSender>,
-) -> BoxedFilter<(impl Reply,)> {
-    self::path()
-        .and(warp::post())
-        .and(has_permission(ROUTE_ADD_PEER, public_routes, allowed_ips))
-        .and(warp::body::json())
-        .and(with_peer_manager(peer_manager))
-        .and(with_network_command_sender(network_command_sender))
-        .and_then(
-            |value, peer_manager, network_controller| async move { add_peer(value, peer_manager, network_controller) },
-        )
-        .boxed()
+pub(crate) fn filter<B: StorageBackend>() -> Router {
+    Router::new().route("/peers", post(add_peer::<B>))
 }
 
-pub(crate) fn add_peer(
-    value: JsonValue,
-    peer_manager: ResourceHandle<PeerManager>,
-    network_controller: ResourceHandle<NetworkCommandSender>,
-) -> Result<impl Reply, Rejection> {
+pub(crate) async fn add_peer<B: StorageBackend>(
+    Json(value): Json<Value>,
+    Extension(args): Extension<ApiArgsFullNode<B>>,
+) -> Result<impl IntoResponse, ApiError> {
     let multi_address_v = &value["multiAddress"];
     let alias_v = &value["alias"];
 
     let mut multi_address = multi_address_v
         .as_str()
-        .ok_or_else(|| {
-            reject::custom(CustomRejection::BadRequest(
-                "invalid multi address: expected a string".to_string(),
-            ))
-        })?
+        .ok_or_else(|| ApiError::BadRequest("invalid multiaddress".to_string()))?
```

Comment:
I feel like the `ApiError` variants should contain `'static str`s instead of `String`s. This way you can prevent those closures and make the code overall a bit nicer.

Reply:
I think static string slices are optimized fancily but I don't have any hard evidence of it

Reply:
But how to solve these kinds of errors?

`.map_err(|e| ApiError::BadRequest(e.to_string()))?;`

If adding `&` infront then the borrow doesn't last for `'static`
Or how would you solve this?

Reply:
You could introduce different error variants, maybe: `ApiError::BadJsonRequest(JsonError)` that simply wraps the error (no extra allocation), and `ApiError::BadRequest(&'static str)` (also no allocation), but if you feel is too much of a burden right now, then just leave it bee. I don't think it's a biggie, but something that can be optimized. Your choice! :slightly_smiling_face:

Reply:
or would something like that work for you?
```rust
#[derive(thiserror::Error, Debug)]
enum ApiError {
   ...
   [#error("{0}")]
   BadRequest(JsonError),
   [#error("Invalid parents: ...")]
   InvalidParentsBadRequest,
   ...
}
```

Reply:
Fixed this